sensors/movementsensor.go

```diff
@@ -0,0 +1,324 @@
+		readingTimePosition, readingTimeOrientation time.Time
+		angVel                                      spatialmath.AngularVelocity
+		linAcc                                      r3.Vector
+		position                                    *geo.Point
+		orientation                                 spatialmath.Orientation
+		timedIMUReadingResponse                     *TimedIMUReadingResponse
+		timedOdometerReadingResponse                *TimedOdometerReadingResponse
+		err                                         error
+	)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, timedMovementSensorReadingTimeout)
+	defer cancel()
+	if ms.imuSupported {
+	imuLoop:
+		for {
+			select {
+			case <-timeoutCtx.Done():
+				return TimedMovementSensorReadingResponse{}, timeoutCtx.Err()
+			default:
+				if timedIMUReadingResponse, err = ms.timedIMUReading(timeoutCtx, &angVel, &linAcc,
+					&readingTimeAngularVel, &readingTimeLinearAcc); err != nil && !errors.Is(err, ErrNoValidReadingObtained) {
+					return TimedMovementSensorReadingResponse{}, err
+				}
+				if timedIMUReadingResponse != nil {
+					break imuLoop
+				}
+			}
+		}
+	}
+	if ms.odometerSupported {
+	odometerLoop:
+		for {
+			select {
+			case <-timeoutCtx.Done():
+				return TimedMovementSensorReadingResponse{}, timeoutCtx.Err()
+			default:
+				if timedOdometerReadingResponse, err = ms.timedOdometerReading(timeoutCtx, position, &orientation,
+					&readingTimePosition, &readingTimeOrientation); err != nil && !errors.Is(err, ErrNoValidReadingObtained) {
+					return TimedMovementSensorReadingResponse{}, err
+				}
+				if timedOdometerReadingResponse != nil {
+					break odometerLoop
+				}
+			}
+		}
+	}
+	return TimedMovementSensorReadingResponse{
+		TimedIMUResponse:      timedIMUReadingResponse,
+		TimedOdometerResponse: timedOdometerReadingResponse,
+		IsReplaySensor:        ms.replay,
+	}, nil
+}
+
+func (ms *MovementSensor) timedIMUReading(ctx context.Context, angVel *spatialmath.AngularVelocity, linAcc *r3.Vector,
+	readingTimeAngularVel, readingTimeLinearAcc *time.Time,
+) (*TimedIMUReadingResponse, error) {
+	var err error
+
+	returnReadingIfTimestampsWithinTolerance := func(readingTimeAngularVel, readingTimeLinearAcc time.Time,
+		angVel *spatialmath.AngularVelocity, linAcc *r3.Vector,
+	) (TimedIMUReadingResponse, bool) {
+		if readingTimeAngularVel.Sub(readingTimeLinearAcc).Abs().Milliseconds() < replayTimeToleranceMsec {
+			return TimedIMUReadingResponse{
+				LinearAcceleration: *linAcc,
+				AngularVelocity: spatialmath.AngularVelocity{
+					X: rdkutils.DegToRad(angVel.X),
+					Y: rdkutils.DegToRad(angVel.Y),
+					Z: rdkutils.DegToRad(angVel.Z),
+				},
+				ReadingTime: readingTimeLinearAcc.Add((readingTimeLinearAcc.Sub(readingTimeAngularVel)).Abs() / 2),
+			}, true
+		}
+		return TimedIMUReadingResponse{}, false
+	}
+
+	if *readingTimeLinearAcc == defaultTime || readingTimeLinearAcc.Sub(*readingTimeAngularVel).Milliseconds() < 0 {
```

Comment:
Can we turn this section into a for loop over LinearAccleration and Angular velocity instead of repeating the code twice? It would make it more digestible.

Reply:
What do you mean by "repeating the code twice"?

Reply:
Agreed, after talking another look I think the added complexity to achieve this wouldn't be beneficial